apps/src/applab/commands.js

```diff
@@ -1212,6 +1226,27 @@ applabCommands.onEventFired = function (opts, e) {
           applabEvent[prop] = (e[prop] - yOffset) / yScale;
         }
       });
+    // Set movementX and movementY, computing it from clientX and clientY if necessary.
+    // The element must have an element id for this to work.
+    if (typeof e.movementX !== 'undefined' && typeof e.movementY !== 'undefined') {
+      // The browser supports movementX and movementY natively.
+      applabEvent.movementX = e.movementX;
+      applabEvent.movementY = e.movementY;
+    } else if (e.type == 'mousemove') {
+      var currentTargetId = e.currentTarget.id;
+      var lastEvent = lastMouseMoveEventMap[currentTargetId];
+      if (currentTargetId && lastEvent) {
+        // Compute movementX and movementY from clientX and clientY.
+        applabEvent.movementX = e.clientX - lastEvent.clientX;
+        applabEvent.movementY = e.clientY - lastEvent.clientY;
+      } else {
+        // There has been no mousemove event on this element since the most recent
+        // mouseout event, or this element does not have an element id.
+        applabEvent.movementX = 0;
+        applabEvent.movementY = 0;
+      }
+      lastMouseMoveEventMap[currentTargetId] = e;
```

Comment:
Correct, however under normal circumstances I would expect a mousemove event at 0,0 prior to the mousedown event at 0,0, since the user had to move the mouse there before they clicked there, though it's possible this would not be the case if their mouse is over the canvas when they load an app share url or switch tabs. Furthermore, we don't give you the movementX/Y on the first mousemove event on an element. These are both shortcomings of this implementation when compared with browsers like Chrome which implement movementX/Y natively.